### DISCUSSION TITLE
The help display is not helpful for prefixes with lots of bindings (like *root-map*) C-t ?

### DISCUSSION THREAD

**Discussion by equwal** [stumpwm/stumpwm#1106] [Ideas]
I don't think the current display is very helpful. Here is a design that I think has some merit, since it groups things into categories instead of just putting them there. Even alphabetizing could help. I see a few different options here:

- Allow for providing a "category" when defining a command so they can come up pretty in the help buffer.
- Alphabetize by command name
- Alphabetize by keybinding
- Group somehow by keybinding (I believe this is the concept behind the current one).
- Maybe do more than one of these and provide a key to cycle the viewing styles (I think this is overly complex).

Personally I am partial to grouping based on categories, but this comes at the cost of arbitrary complexity when defining commands. This can be with more versatility and without any sort of breaking changes by adding a value `category` to the plist of any command like this:
```lisp
(progn (setf (get 'emacs 'category) 'applications) 
       (get 'emacs 'category))
```
Also, the commands are already split into maps anyway in `bindings.lisp` such as `*groups-map*`, `*exchange-window-map*`, etc.

https://lh6.googleusercontent.com/2HE0ct4UkSFsQEm5A-e2nPxi4Wn2a14elnmovolM9xYhIXItjyVMgKjvAfo7AlZ0laSrFGwi2r22br9yfi9U=w1598-h805

[reddit post by that designer](https://www.reddit.com/r/stumpwm/comments/e1r6hp/just_throwing_this_out_here/)

Another point is that it isn't obvious how to get more documentation on a given keybinding. If you don't know what `banish` does, then you must go find and open the docs. But banish has a docstring, so executing:
`C-t ? k C-t b` should show that, rather than merely `C-i b is bound to banish`. This concern is resolved in #671, though it would be nice to view a summary of the commands from the buffer also.

**Comment by dmb2** [comment]
Thanks for the additional context of the reddit post, without that I would be confused.

I agree that this is something that needs work, but as you have found out, it is a little more nuanced than it first appears.  The help map that comes up must be dynamic so that it reflects StumpWM's current state, not the default state.  

I can't see the link you posted, but I looked at the pictures in the reddit post, I like the concept, but I'm not sure how to make it scalable and dynamic.  Maybe instead of tagging a property we organize them into different maps and have those flow?  

Another thing I've always been a bit annoyed about is how the help menu shows exactly what string is bound, so if you have something long and complicated, it will make your menu much wider than necessary if it just put an ellipses after 10-20 characters.

Are you working with the reddit post author? It looks like you two have some good collaboration going, and while using a static string isn't going to be our final solution, I like the way the headings are formatted, its much easier to read.

**Comment by equwal** [comment]
> I agree that this is something that needs work, but as you have found out, it is a little more nuanced than it first appears. The help map that comes up must be dynamic so that it reflects StumpWM's current state, not the default state.

Indeed, I pointed this out to that author (my reddit handle is `bjyo`).

> I can't see the link you posted, but I looked at the pictures in the reddit post, I like the concept, but I'm not sure how to make it scalable and dynamic. Maybe instead of tagging a property we organize them into different maps and have those flow?

That is almost what is going on in `bindings.lisp`, so I think relatively few changes need to be made to group those bindings. The hard work is getting everything showing up properly and sized right. Just adding a few more maps, and creating this "flow" could go a long way.

> Another thing I've always been a bit annoyed about is how the help menu shows exactly what string is bound, so if you have something long and complicated, it will make your menu much wider than necessary if it just put an ellipses after 10-20 characters.

Maybe provide a mechanism for abbreviating command names explicitly.

> Are you working with the reddit post author? It looks like you two have some good collaboration going, and while using a static string isn't going to be our final solution, I like the way the headings are formatted, its much easier to read.

I am not the reddit post author. I'd like him to come work here, but if he doesn't it shouldn't matter too much. I referenced him because I thought the concept behind his screenshots was nice and reasonably doable.

**Comment by Vorlonhomeworld** [comment]
I'm the reditor in question. The code is definetly nothing fancy as I just started learning lisp when I installed Stumpwm

https://gist.github.com/57e3c3a40a9e20b5b2ad0564871453ec

I'm basically using the "message" function in Stumpwm to display the menu.  Yes, it's bigger than I intended, however I'm not yet proficient in Common Lisp to update it yet.

**Comment by dmb2** [comment]
Welcome @Vorlonhomeworld! We're excited to have you contributing to StumpWM!

I really like the concept of what you've started to outline, but we've got to figure out a way to make it happen automatically instead of it being a big static string the way you've defined it.  I love the typographic formatting that you've chosen, but if we could get that by iterating over the various key-binding maps and printing strings that would be great. 

The code to produce a string like the one you've got is going to be pretty complicated, but well worth it.  Feel free to ask questions here as you progress.  Maybe you can coordinate with @equwal since he's already got a good idea of what needs to be done from an implementation side of things.

**Comment by Vorlonhomeworld** [comment]
I agree @dmb2. I knew nothing about Common Lisp when I started this, I'm learning and working on a way to have the menu take less real estate on the screen. The. Stumpwmrc is fairly easy to do, but there are two other files with commands in them that I may not be able to work with in the same way, but I'll give it a try anyway!

**Comment by equwal** [comment]
https://github.com/stumpwm/stumpwm/pull/671
Closing this now I guess.
